staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"k8s.io/klog/v2"
+	"k8s.io/klog/v2/ktesting"
+	"k8s.io/utils/clock"
+	testingclock "k8s.io/utils/clock/testing"
+)
+
+func TestConditionalProgressRequester(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	logger := klog.FromContext(ctx)
+
+	clock := testingclock.NewFakeClock(time.Now())
+	pr := newTestConditionalProgressRequester(clock)
+	stopCh := make(chan struct{})
+	go pr.Run(stopCh)
+	var wantRequestsSent int32
+
+	logger.Info("No progress requests if no-one is waiting")
+	clock.Step(progressRequestPeriod * 2)
+	time.Sleep(time.Millisecond)
```

Comment:
Didn't find a good way to ensure background goroutine runs fully. runtime.Gosched() is flaky :(

Reply:
I can guarantee that time.Sleep(1ms) will be flaky (our env is too overloaded to make it work reliably).

The only reliable way to make it work is sth like:
```
if err := wait.Poll(1ms, 1s?, func() { if pr.progressRequestsSentCount.Load() != ... { ... }}); err != nil {
   ...
}
```

[so basically polling]